operator/video/v0/video_operation.go

```diff
@@ -0,0 +1,195 @@
+	}
+
+	tempInputFileName := tempInputFile.Name()
+	defer os.Remove(tempInputFileName)
+
+	if _, err := tempInputFile.Write(videoBytes); err != nil {
+		return nil, fmt.Errorf("error in writing file: %s", err)
+	}
+
+	split := ffmpeg.Input(tempInputFileName)
+
+	tempOutputFile, err := os.CreateTemp("", "temp_out.*.mp4")
+	if err != nil {
+		return nil, fmt.Errorf("error in creating temp output file: %s", err)
+	}
+	tempOutputFileName := tempOutputFile.Name()
+	defer os.Remove(tempOutputFileName)
+
+	split = split.OverWriteOutput()
+	err = split.Output(tempOutputFileName, getKwArgs(inputStruct)).Run()
+
+	if err != nil {
+		return nil, fmt.Errorf("error in running ffmpeg: %s", err)
+	}
+
+	byOut, _ := os.ReadFile(tempOutputFileName)
+	base64Subsample := videoPrefix + "," + base64.StdEncoding.EncodeToString(byOut)
+
+	output := SubsampleVideoOutput{
+		Video: Video(base64Subsample),
+	}
+
+	return base.ConvertToStructpb(output)
+}
+
+func getKwArgs(inputStruct SubsampleVideoInput) ffmpeg.KwArgs {
+	kwArgs := ffmpeg.KwArgs{"r": inputStruct.Fps}
+	if inputStruct.StartTime != "" {
+		kwArgs["ss"] = inputStruct.StartTime
+	}
+	if inputStruct.Duration != "" {
+		kwArgs["t"] = inputStruct.Duration
+	}
+	return kwArgs
+}
+
+func subsampleVideoFrames(input *structpb.Struct) (*structpb.Struct, error) {
+	inputStruct := SubsampleVideoFramesInput{}
+
+	err := base.ConvertFromStructpb(input, &inputStruct)
+	if err != nil {
+		return nil, fmt.Errorf("error converting input to struct: %v", err)
+	}
+
+	base64Video := string(inputStruct.Video)
+
+	videoBytes, err := base64.StdEncoding.DecodeString(base.TrimBase64Mime(base64Video))
+
+	if err != nil {
+		return nil, fmt.Errorf("error in decoding for inner: %s", err)
+	}
+
+	tempInputFile, err := os.CreateTemp("", "temp.*.mp4")
+	if err != nil {
+		return nil, fmt.Errorf("error in creating temp input file: %s", err)
+	}
+	tempInputFileName := tempInputFile.Name()
+	defer os.Remove(tempInputFileName)
+
+	if _, err := tempInputFile.Write(videoBytes); err != nil {
+		return nil, fmt.Errorf("error in writing file: %s", err)
+	}
+
+	random := uuid.New().String()
+	// TODO: chuang8511 confirm the reasonable numbers for outputPattern.
+	// In the future, we will support bigger size of video, so we set the frame number to 8 digits.
+	// Because the sequence is important, we need to use pattern
+	// with frame number rather than uuid as suffix.
+	outputPattern := random + "_frame_%08d.jpeg"
+
+	err = ffmpeg.Input(tempInputFileName).
+		Output(outputPattern,
+			getFramesKwArgs(inputStruct),
+		).
+		Run()
+
+	if err != nil {
+		return nil, fmt.Errorf("error in running ffmpeg: %s", err)
+	}
+
+	files, err := filepath.Glob(random + "_frame_*.jpeg")
+	if err != nil {
+		return nil, fmt.Errorf("error listing frames: %s", err)
+	}
+
+	sort.Strings(files)
+	jpegPrefix := "data:image/jpeg;base64,"
+	var frames []Frame
+	for _, file := range files {
+		defer os.Remove(file)
```

Comment:
You should place the `defer` statement before the loop. Otherwise, if the loop returns early, some images won't be deleted.

Reply:
modified!